### PR TITLE
Adds support for API Registry and enable data residency

### PR DESCRIPTION
**Pull request by srinandan**

### CODE REVIEW

File: cmd/apis/bundlecrtapis.go [hunk 1]
```diff
@@ -110,7 +111,7 @@ func init() {
 
 	BundleCreateCmd.Flags().StringVarP(&env, "env", "e",
 		"", "Name of the environment to deploy the proxy")
-	BundleCreateCmd.Flags().BoolVarP(&overrides, "ovr", "r",
+	BundleCreateCmd.Flags().BoolVarP(&overrides, "ovr", "",
```
Comment (ssvaidyanathan) [inline cmd/apis/bundlecrtapis.go:114]: wont this have issues? What if existing users are using -r to pass this argument?

File: cmd/apis/depapi.go [hunk 2]
```diff
@@ -68,7 +69,7 @@ func init() {
 		"", "Apigee environment name")
 	DepCmd.Flags().IntVarP(&revision, "rev", "v",
 		-1, "API Proxy revision. If not set, the highest revision is used")
-	DepCmd.Flags().BoolVarP(&overrides, "ovr", "r",
+	DepCmd.Flags().BoolVarP(&overrides, "ovr", "",
```
Comment (ssvaidyanathan) [inline cmd/apis/depapi.go:72]: same comment as previous one

File: cmd/apis/ghcrtapis.go [hunk 3]
```diff
@@ -67,7 +67,7 @@ func init() {
 		"", "API Proxy name")
 	GhCreateCmd.Flags().StringVarP(&ghOwner, "owner", "u",
 		"", "The github organization or username. ex: In https://github.com/apigee, apigee is the owner name")
-	GhCreateCmd.Flags().StringVarP(&ghRepo, "repo", "r",
+	GhCreateCmd.Flags().StringVarP(&ghRepo, "repo", "",
```
Comment (ssvaidyanathan) [inline cmd/apis/ghcrtapis.go:70]: same comment as previous one

File: cmd/apis/listdeploy.go [hunk 4]
```diff
@@ -67,7 +68,7 @@ func init() {
 		"", "API proxy name")
 	ListDepCmd.Flags().StringVarP(&env, "env", "e",
 		"", "Apigee environment name")
-	ListDepCmd.Flags().IntVarP(&revision, "rev", "r",
+	ListDepCmd.Flags().IntVarP(&revision, "rev", "v",
```
Comment (ssvaidyanathan) [inline cmd/apis/listdeploy.go:71]: changing it from `r` to `v`.. Wont this break for users who are using `r`?

File: cmd/appgroups/createkey.go [hunk 5]
```diff
@@ -55,7 +56,7 @@ func init() {
 		"", "Name of the app")
 	CreateKeyCmd.Flags().StringVarP(&key, "key", "k",
 		"", "Import an existing AppGroup app consumer key")
-	CreateKeyCmd.Flags().StringVarP(&secret, "secret", "r",
+	CreateKeyCmd.Flags().StringVarP(&secret, "secret", "c",
```
Comment (ssvaidyanathan) [inline cmd/appgroups/createkey.go:59]: change not backward compatible as the arg is changing

File: cmd/apps/createkey.go [hunk 6]
```diff
@@ -47,7 +48,7 @@ var CreateKeyCmd = &cobra.Command{
 func init() {
 	CreateKeyCmd.Flags().StringVarP(&key, "key", "k",
 		"", "Developer app consumer key")
-	CreateKeyCmd.Flags().StringVarP(&secret, "secret", "r",
+	CreateKeyCmd.Flags().StringVarP(&secret, "secret", "c",
```
Comment (ssvaidyanathan) [inline cmd/apps/createkey.go:51]: change not backward compatible as the arg is changing

File: cmd/env/crttraceoverrides.go [hunk 7]
```diff
@@ -48,7 +49,7 @@ func init() {
 		"", "Trace endpoint, used only with JAEGER")
 	CrtTraceOverridesCmd.Flags().StringVarP(&sampler, "sampler", "s",
 		"PROBABILITY", "Sampler can be set to PROBABILITY or OFF")
-	CrtTraceOverridesCmd.Flags().StringVarP(&sampleRate, "rate", "r",
+	CrtTraceOverridesCmd.Flags().StringVarP(&sampleRate, "rate", "",
```
Comment (ssvaidyanathan) [inline cmd/env/crttraceoverrides.go:52]: why was this removed?

File: cmd/env/removerole.go [hunk 8]
```diff
@@ -45,7 +45,7 @@ var RemoveRoleCmd = &cobra.Command{
 func init() {
 	RemoveRoleCmd.Flags().StringVarP(&memberName, "name", "n",
 		"", "Member name. Prefix the member role as role:name. Ex: user:foo@acme.com")
-	RemoveRoleCmd.Flags().StringVarP(&role, "role", "r",
+	RemoveRoleCmd.Flags().StringVarP(&role, "role", "",
```
Comment (ssvaidyanathan) [inline cmd/env/removerole.go:48]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `role` is breaking change right?

File: cmd/env/setax.go [hunk 9]
```diff
@@ -52,7 +52,7 @@ func init() {
 		"", "Member Name, example Service Account Name")
 	SetAxCmd.Flags().StringVarP(&memberType, "memberType", "m",
 		"serviceAccount", "memberType must be serviceAccount, user or group")
-	SetAxCmd.Flags().StringVarP(&role, "memberRole", "r",
+	SetAxCmd.Flags().StringVarP(&role, "memberRole", "",
```
Comment (ssvaidyanathan) [inline cmd/env/setax.go:55]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `role` is breaking change right?

File: cmd/env/setcustom.go [hunk 10]
```diff
@@ -53,7 +53,7 @@ var SetCustCmd = &cobra.Command{
 func init() {
 	SetCustCmd.Flags().StringVarP(&memberName, "name", "n",
 		"", "Member Name, example Service Account Name")
-	SetCustCmd.Flags().StringVarP(&role, "role", "r",
+	SetCustCmd.Flags().StringVarP(&role, "role", "",
```
Comment (ssvaidyanathan) [inline cmd/env/setcustom.go:56]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `role` is breaking change right?

File: cmd/env/testiam.go [hunk 11]
```diff
@@ -42,6 +42,6 @@ var verb, resource string
 func init() {
 	TestIamCmd.Flags().StringVarP(&verb, "verb", "v",
 		"get", "resource verb")
-	TestIamCmd.Flags().StringVarP(&resource, "res", "r",
+	TestIamCmd.Flags().StringVarP(&resource, "res", "s",
```
Comment (ssvaidyanathan) [inline cmd/env/testiam.go:45]: not backward compatible

File: cmd/env/updatetracecfg.go [hunk 12]
```diff
@@ -46,7 +46,7 @@ func init() {
 		"", "Trace endpoint, used only with JAEGER")
 	UpdateTraceConfigCmd.Flags().StringVarP(&sampler, "sampler", "s",
 		"PROBABILITY", "Sampler can be set to PROBABILITY or OFF")
-	UpdateTraceConfigCmd.Flags().StringVarP(&sampleRate, "rate", "r",
+	UpdateTraceConfigCmd.Flags().StringVarP(&sampleRate, "rate", "",
```
Comment (ssvaidyanathan) [inline cmd/env/updatetracecfg.go:49]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `rate` is breaking change right?

File: cmd/iam/create.go [hunk 13]
```diff
@@ -53,7 +53,7 @@ func init() {
 		"", "Service Account Name")
 	CallCmd.Flags().BoolVarP(&generateName, "gen", "g",
 		false, "Generate account name")
-	CallCmd.Flags().StringVarP(&roleType, "role", "r",
+	CallCmd.Flags().StringVarP(&roleType, "role", "",
```
Comment (ssvaidyanathan) [inline cmd/iam/create.go:56]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `role` is breaking change right?

File: cmd/org/createorg.go [hunk 14]
```diff
@@ -66,7 +66,7 @@ var (
 func init() {
 	CreateCmd.Flags().StringVarP(&description, "desc", "d",
 		"", "Apigee org description")
-	CreateCmd.Flags().StringVarP(&analyticsRegion, "reg", "r",
+	CreateCmd.Flags().StringVarP(&analyticsRegion, "reg", "",
```
Comment (ssvaidyanathan) [inline cmd/org/createorg.go:69]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `reg` is breaking change right?

File: cmd/org/delorg.go [hunk 15]
```diff
@@ -41,6 +41,6 @@ var retention string
 func init() {
 	DelCmd.Flags().StringVarP(&org, "org", "o",
 		"", "Apigee organization name")
-	DelCmd.Flags().StringVarP(&retention, "retension", "r",
+	DelCmd.Flags().StringVarP(&retention, "retension", "",
```
Comment (ssvaidyanathan) [inline cmd/org/delorg.go:44]: not backward compatible right?

File: cmd/products/delrateplan.go [hunk 16]
```diff
@@ -39,7 +40,7 @@ var DelRatePlanCmd = &cobra.Command{
 func init() {
 	DelRatePlanCmd.Flags().StringVarP(&name, "name", "n",
 		"", "name of the API Product")
-	DelRatePlanCmd.Flags().StringVarP(&rateplan, "rateplan", "r",
+	DelRatePlanCmd.Flags().StringVarP(&rateplan, "rateplan", "p",
```
Comment (ssvaidyanathan) [inline cmd/products/delrateplan.go:43]: not backward compatible

File: cmd/products/getrateplan.go [hunk 17]
```diff
@@ -39,7 +40,7 @@ var GetRatePlanCmd = &cobra.Command{
 func init() {
 	GetRatePlanCmd.Flags().StringVarP(&name, "name", "n",
 		"", "name of the API Product")
-	GetRatePlanCmd.Flags().StringVarP(&rateplan, "rateplan", "r",
+	GetRatePlanCmd.Flags().StringVarP(&rateplan, "rateplan", "p",
```
Comment (ssvaidyanathan) [inline cmd/products/getrateplan.go:43]: not backward cmopatible

File: cmd/references/crtref.go [hunk 18]
```diff
@@ -50,7 +50,7 @@ func init() {
 		"", "Description for the reference")
 	CreateCmd.Flags().StringVarP(&resourceType, "restype", "s",
 		"", "Resource type must be KeyStore or TrustStore")
-	CreateCmd.Flags().StringVarP(&refers, "refers", "r",
+	CreateCmd.Flags().StringVarP(&refers, "refers", "",
```
Comment (ssvaidyanathan) [inline cmd/references/crtref.go:53]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `refers` is breaking change right?

File: cmd/references/updateref.go [hunk 19]
```diff
@@ -49,7 +49,7 @@ func init() {
 		"", "Description for the reference")
 	UpdateCmd.Flags().StringVarP(&resourceType, "restype", "s",
 		"", "Resource type must be KeyStore or TrustStore")
-	UpdateCmd.Flags().StringVarP(&refers, "refers", "r",
+	UpdateCmd.Flags().StringVarP(&refers, "refers", "",
```
Comment (ssvaidyanathan) [inline cmd/references/updateref.go:52]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `refers` is breaking change right?

File: cmd/res/crtres.go [hunk 20]
```diff
@@ -44,7 +45,7 @@ func init() {
 		"", "Name of the resource file")
 	CreateCmd.Flags().StringVarP(&resType, "type", "p",
 		"", "Resource type; Valid types include java, js, jsc, properties, py, wsdl, xsd, or xsl.")
-	CreateCmd.Flags().StringVarP(&resPath, "respath", "r",
+	CreateCmd.Flags().StringVarP(&resPath, "respath", "",
```
Comment (ssvaidyanathan) [inline cmd/res/crtres.go:48]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `respath` is breaking change right?

File: cmd/res/updres.go [hunk 21]
```diff
@@ -41,7 +42,7 @@ func init() {
 		"", "Name of the resource file")
 	UpdateCmd.Flags().StringVarP(&resType, "type", "p",
 		"", "Resource type. Valid types include java, js, jsc, properties, py, wsdl, xsd, or xsl.")
-	UpdateCmd.Flags().StringVarP(&resPath, "respath", "r",
+	UpdateCmd.Flags().StringVarP(&resPath, "respath", "",
```
Comment (ssvaidyanathan) [inline cmd/res/updres.go:45]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `respath` is breaking change right?

File: cmd/securityprofiles/attach.go [hunk 22]
```diff
@@ -43,7 +44,7 @@ func init() {
 		"", "Name of the security profile")
 	AttachCmd.Flags().StringVarP(&environment, "env", "e",
 		"", "Apigee environment name")
-	AttachCmd.Flags().StringVarP(&revision, "rev", "r",
+	AttachCmd.Flags().StringVarP(&revision, "rev", "v",
```
Comment (ssvaidyanathan) [inline cmd/securityprofiles/attach.go:47]: not backward compatible

File: cmd/securityprofiles/get.go [hunk 23]
```diff
@@ -40,7 +41,7 @@ var name string
 func init() {
 	GetCmd.Flags().StringVarP(&name, "name", "n",
 		"", "Name of the security profile")
-	GetCmd.Flags().StringVarP(&revision, "revision", "r",
+	GetCmd.Flags().StringVarP(&revision, "revision", "v",
```
Comment (ssvaidyanathan) [inline cmd/securityprofiles/get.go:44]: not backward compatible

File: cmd/sharedflows/bundlecrtsf.go [hunk 24]
```diff
@@ -109,7 +110,7 @@ func init() {
 		"", "Path to the Sharedflow Bundle; ex: ./test/sharedflowbundle")
 	BundleCreateCmd.Flags().StringVarP(&env, "env", "e",
 		"", "Apigee environment name")
-	BundleCreateCmd.Flags().BoolVarP(&overrides, "ovr", "r",
+	BundleCreateCmd.Flags().BoolVarP(&overrides, "ovr", "",
```
Comment (ssvaidyanathan) [inline cmd/sharedflows/bundlecrtsf.go:113]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `ovr` is breaking change right?

File: cmd/sharedflows/depsf.go [hunk 25]
```diff
@@ -62,7 +63,7 @@ func init() {
 		"", "Apigee environment name")
 	DepCmd.Flags().IntVarP(&revision, "rev", "v",
 		-1, "Sharedflow revision. If not set, the highest revision is used")
-	DepCmd.Flags().BoolVarP(&overrides, "ovr", "r",
+	DepCmd.Flags().BoolVarP(&overrides, "ovr", "",
```
Comment (ssvaidyanathan) [inline cmd/sharedflows/depsf.go:66]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `ovr` is breaking change right?

File: cmd/sharedflows/ghcrtsf.go [hunk 26]
```diff
@@ -67,7 +67,7 @@ func init() {
 		"", "Sharedflow name")
 	GhCreateCmd.Flags().StringVarP(&ghOwner, "owner", "u",
 		"", "The github organization or username. ex: In https://github.com/apigee, apigee is the owner name")
-	GhCreateCmd.Flags().StringVarP(&ghRepo, "repo", "r",
+	GhCreateCmd.Flags().StringVarP(&ghRepo, "repo", "",
```
Comment (ssvaidyanathan) [inline cmd/sharedflows/ghcrtsf.go:70]: why was this removed? Is it because of the global `r` flag? If yes, then removing the existing `r` for `repo` is breaking change right?

File: cmd/sharedflows/listdeploy.go [hunk 27]
```diff
@@ -61,6 +62,6 @@ func init() {
 		"", "Shareflow name")
 	ListDepCmd.Flags().StringVarP(&env, "env", "e",
 		"", "Apigee environment name")
-	ListDepCmd.Flags().IntVarP(&revision, "rev", "r",
+	ListDepCmd.Flags().IntVarP(&revision, "rev", "v",
```
Comment (ssvaidyanathan) [inline cmd/sharedflows/listdeploy.go:65]: not backward compatible

File: cmd/sharedflows/listsf.go [hunk 28]
```diff
@@ -42,6 +43,6 @@ var includeRevisions bool
 func init() {
 	ListCmd.Flags().StringVarP(&env, "env", "e",
 		"", "Apigee environment name")
-	ListCmd.Flags().BoolVarP(&includeRevisions, "rev", "r",
+	ListCmd.Flags().BoolVarP(&includeRevisions, "rev", "v",
```
Comment (ssvaidyanathan) [inline cmd/sharedflows/listsf.go:46]: not backward compatible

File: internal/client/registry/apis/versions/artifacts.go [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+	return respBody, err
+}
+
+// ListArtifacts
+func ListArtifacts(apiName string, apiVersion string, pageSize int,
+	pageToken string, filter string, orderBy string,
+) (respBody []byte, err error) {
+	u, _ := url.Parse(apiclient.GetApigeeRegistryURL())
+	u.Path = path.Join(u.Path, "apis", apiName, "versions", apiVersion, "artifacts")
+	return common.ListArtifacts(u, pageSize, pageToken, filter, orderBy)
```
Comment (kurtkanaskie) [inline internal/client/registry/apis/versions/artifacts.go:76]: I see this error here:
```
gorun registry apis versions artifacts list --org=$ORG --api-name=ui-api-id-1
{
  "error": {
    "code": 400,
    "message": "Invalid JSON payload received. Unknown name \"pageSize\": Cannot bind query parameter. Field 'pageSize' could not be found in request message.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.BadRequest",
        "fieldViolations": [
          {
            "description": "Invalid JSON payload received. Unknown name \"pageSize\": Cannot bind query parameter. Field 'pageSize' could not be found in request message."
          }
        ]
      }
    ]
  }
}
```

**Review by kurtkanaskie — CHANGES_REQUESTED** [review]
OK, I think I figured out how to navigate the APIs, versions to get to the spec, but we are missing getContent for the spec
```
gorun registry apis versions artifacts specs 
Manage specs for an API version in Apigee Registry

Usage:
  apigeecli registry apis versions artifacts specs [command]

Available Commands:
  list        List all specs for an API version
```
The equivalent of this API
```
curlx https://apigeeregistry.googleapis.com/v1/projects/apigeex-mint-kurt/locations/global/apis/ui-api-id-1/versions/v1/specs/petstore-30yaml:getContents
```
I think we need:
`gorun registry apis versions specs get content`
Along with
`gorun registry apis versions specs create`
`gorun registry apis versions specs get`
`gorun registry apis versions specs delete`
`gorun registry apis versions specs update`

I don't think it should be under "artifacts" as it is now
`apigeecli registry apis versions artifacts specs`
just versions, artifacts are something else.
`apigeecli registry apis versions specs -h`

File: internal/client/registry/apis/versions/artifacts.go (hunk 29 above)
Comment (srinandan) [reply internal/client/registry/apis/versions/artifacts.go:76]: The command is incorrect. I have added validations now. `api-name` and `api-version` are mandatory parameters.

**Comment by srinandan** [comment]
The correct command for what you are trying is: 

```
apigeecli registry apis versions artifacts specs get --help
Get the spec details for an API version in Apigee Registry

Usage:
  apigeecli registry apis versions artifacts specs get [flags]

Flags:
      --api-name string      API Name
      --api-version string   API Version
  -c, --content              If set to true, returns artifact contents
  -h, --help                 help for get
  -n, --name string          Name of the Spec
```

File: cmd/registry/apis/versions/specs/getspec.go [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+	Long:  "Get the spec details for an API version in Apigee Registry",
+	Args: func(cmd *cobra.Command, args []string) (err error) {
+		apiclient.SetProjectID(utils.ProjectID)
+		apiclient.SetRegion(utils.Region)
+		return apiclient.SetApigeeOrg(utils.Org)
+	},
+	RunE: func(cmd *cobra.Command, args []string) (err error) {
+		if content {
+			_, err = versions.GetSpecContents(apiName, apiVersion, name)
+			return err
```
Comment (kurtkanaskie) [inline cmd/registry/apis/versions/specs/getspec.go:39]: Should just be return

**Review by kurtkanaskie — CHANGES_REQUESTED** [review]
I don't think the spec content should be associated in the context of an artifact.
Structure of APIs is:
projects.locations
projects.locations.apis
projects.locations.apis.artifacts # Related, Dependencies
projects.locations.apis.deployments
projects.locations.apis.deployments.artifacts
projects.locations.apis.versions
projects.locations.apis.versions.artifacts
projects.locations.apis.versions.specs:getContents # OAS
projects.locations.apis.versions.specs.artifacts
projects.locations.artifacts

This API returns spec content, notice there's no artifact in the path.
curl https://apigeeregistry.googleapis.com/v1/projects/apigeex-mint-kurt/locations/global/apis/ui-api-id-1/versions/v1/specs/petstore-30yaml:getContents

I think this apigeecli
gorun registry apis versions **artifacts** specs get --org=$ORG --name=petstore-30yaml --api-version=v1 --api-name=ui-api-id-1 --content=true

should not have artifacts, just
gorun registry apis versions specs get --org=$ORG --name=petstore-30yaml --api-version=v1 --api-name=ui-api-id-1 --content=true

**Comment by kurtkanaskie** [comment]
Getting content from an API spec version should output to the named filename
```
gorun registry apis versions specs list --org=$ORG --api-name=ui-api-id-1 --api-version=v1
{
  "apiSpecs": [
    {
      "name": "projects/apigeex-mint-kurt/locations/global/apis/ui-api-id-1/versions/v1/specs/petstore-30yaml",
      "filename": "petstore-3.0.yaml",
      "revisionId": "86d6e250",
      "createTime": "2024-01-04T20:51:39.885140Z",
      "revisionCreateTime": "2024-01-04T20:51:39.885140Z",
      "revisionUpdateTime": "2024-01-04T20:51:39.885140Z",
      "mimeType": "application/x.openapi",
      "sizeBytes": 21273,
      "hash": "479e90bfc004ca0386996f0ecda217810e955d3ed4d84a56a2f6e098a158215a"
    }
  ]
}

gorun registry apis versions specs get --org=$ORG --name=petstore-30yaml --api-version=v1 --api-name=ui-api-id-1 --content
```
Creates file: petstore-30yaml.txt but should use filename: petstore-3.0.yaml

**Review by kurtkanaskie — APPROVED** [review]

**Comment by kurtkanaskie** [comment]
LGTM!
